Mathlib/Data/Nat/Fib/Zeckendorf.lean

```diff
@@ -0,0 +1,177 @@
+  (l.map fib).sum < fib n
+| n, [], _, hn => fib_pos.2 $ hn _ rfl
+| n, a :: l, hl, hn => by
+  simp only [IsZeckendorfRep, cons_append, chain'_iff_pairwise, pairwise_cons] at hl
+  have : ∀ b, b ∈ head? (l ++ [0]) → b < a - 1 :=
+    fun b hb ↦ lt_tsub_iff_right.2 $ hl.1 _ $ mem_of_mem_head? hb
+  simp only [mem_append, mem_singleton, ←chain'_iff_pairwise, or_imp, forall_and, forall_eq,
+    zero_add] at hl
+  simp only [map, List.sum_cons]
+  refine (add_lt_add_left (sum_fib_lt hl.2 this) _).trans_le ?_
+  rw [add_comm, ←fib_add_one (hl.1.2.trans_lt' zero_lt_two).ne']
+  exact fib_mono (hn _ rfl)
+
+end List
+
+namespace Nat
+variable {l : List ℕ} {a m n : ℕ}
+
+/-- The greatest index of a Fibonacci number less than or equal to `n`. -/
+def greatestFib (n : ℕ) : ℕ := (n + 1).findGreatest (fun k ↦ fib k ≤ n)
+
+lemma fib_greatestFib_le (n : ℕ) : fib (greatestFib n) ≤ n :=
+findGreatest_spec (P := (fun k ↦ fib k ≤ n)) (zero_le _) $ zero_le _
+
+lemma greatestFib_mono : Monotone greatestFib :=
+fun _a _b hab ↦ findGreatest_mono (fun _k ↦ hab.trans') $ add_le_add_right hab _
+
+@[simp] lemma le_greatestFib : m ≤ greatestFib n ↔ fib m ≤ n :=
+⟨fun h ↦ (fib_mono h).trans $ fib_greatestFib_le _,
+  fun h ↦ le_findGreatest (m.le_fib_add_one.trans $ add_le_add_right h _) h⟩
+
+@[simp] lemma greatestFib_lt : greatestFib m < n ↔ m < fib n :=
+lt_iff_lt_of_le_iff_le le_greatestFib
+
+lemma lt_fib_greatestFib_add_one (n : ℕ) : n < fib (greatestFib n + 1) :=
+greatestFib_lt.1 $ lt_succ_self _
+
+@[simp] lemma greatestFib_fib : ∀ {n}, n ≠ 1 → greatestFib (fib n) = n
+| 0, _ => rfl
+| _n + 2, _ => findGreatest_eq_iff.2
+  ⟨le_fib_add_one _, fun _ ↦ le_rfl, fun _m hnm _ ↦ ((fib_lt_fib le_add_self).2 hnm).not_le⟩
+
+@[simp] lemma greatestFib_eq_zero : greatestFib n = 0 ↔ n = 0 :=
+⟨fun h ↦ by simpa using findGreatest_eq_zero_iff.1 h zero_lt_one le_add_self, by rintro rfl; rfl⟩
+
+lemma greatestFib_ne_zero : greatestFib n ≠ 0 ↔ n ≠ 0 := greatestFib_eq_zero.not
+
+@[simp] lemma greatestFib_pos : 0 < greatestFib n ↔ 0 < n := by simp [pos_iff_ne_zero]
+
+lemma greatestFib_sub_fib_greatestFib_le_greatestFib (hn : n ≠ 0) :
+  greatestFib (n - fib (greatestFib n)) ≤ greatestFib n - 2 := by
+  rw [←lt_succ_iff, greatestFib_lt, tsub_lt_iff_right n.fib_greatestFib_le, Nat.sub_succ, succ_pred,
+    ←fib_add_one]
+  exact n.lt_fib_greatestFib_add_one
+  · simpa
+  · simpa [←succ_le_iff] using hn.bot_lt
+
+private lemma zeckendorf_aux (hm : 0 < m) : m - fib (greatestFib m) < m :=
+tsub_lt_self hm $ fib_pos.2 $ findGreatest_pos.2 ⟨1, zero_lt_one, le_add_self, hm⟩
+
+/-- The Zeckendorf representation of a natural number. -/
+def zeckendorf : ℕ → List ℕ
+| 0 => []
+| m@(n + 1) => let a := greatestFib m
+  have := zeckendorf_aux (n.succ_pos.trans_eq $ Eq.symm ‹_›)
+  a :: zeckendorf (m - fib a)
+
+@[simp] lemma zeckendorf_zero : zeckendorf 0 = [] := rfl
+-- The linter is hallucinating: It must be seeing the `have` from within the def of `zeckendorf`
```

Comment:
I'm not sure it's hallucinating, the have is really there, the issue is that it shouldn't be in the equation lemmas at all. See https://leanprover.zulipchat.com/#narrow/stream/348111-std4/topic/spurious.20.60unusedHavesSuffices.60.20linter.20error.3F one workaround is to but the have inside the `decreasing_by` proof as in https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.236582

Reply:
But I can just nolint it, right? I've written a better equation lemma anyway.

Reply:
The issue is only if anyone else will ever unfold the definition directly yeah. I forget if there is a way to register your equation lemmas as _the_ equation lemmas, there was some discussion about that being possible at some point.

Reply:
Haven't seen that. I've left a comment to the user's attention for now.

Reply:
Yes, that's `@[eqns]`